libcontainer/cgroups/utils.go

```diff
@@ -249,15 +250,24 @@ func RemovePath(path string) error {
 // If after all there are not removed cgroups - appropriate error will be
 // returned.
 func RemovePaths(paths map[string]string) (err error) {
+	const retries = 5
 	delay := 10 * time.Millisecond
-	for i := 0; i < 5; i++ {
+	for i := 0; i < retries; i++ {
 		if i != 0 {
 			time.Sleep(delay)
 			delay *= 2
 		}
 		for s, p := range paths {
-			os.RemoveAll(p)
-			// TODO: here probably should be logging
+			if err := RemovePath(p); err != nil {
+				// do not log intermediate iterations
+				switch i {
+				case 0:
+					logrus.WithError(err).Warnf("Failed to remove cgroup (will retry)")
+				case retries - 1:
+					logrus.WithError(err).Error("Failed to remove cgroup")
```

Comment:
Not sure we need these logrus. The error is returned to the caller.

Reply:
No it's not. The final error returned is `fmt.Errorf("Failed to remove paths: %v", paths)` (see the very end of the function below), it does not say anything about what happened.

This logging addresses the TODO above (`TODO: here probably should be logging`).

I think it makes sense to do what I did -- log the first iteration error as a warning, and last iteration error as an error.

Alternatively, I can try to rework this to return the error to the caller -- the problem is which one to return in case there are a few, during the last iteration. Just pile them up?

Reply:
We could use something like https://github.com/hashicorp/go-multierror to aggregate the errors.

Reply:
So, what this code does is
 - logs the first removal error as a warning (warning because it will retry);
 - retries a few times;
 - if removal still fails, logs the last error as an error.

I believe what the code does is the best approach, let me explain the logic that lead to this code.

1. In case the removal eventually succeeds, we might still be interested in why it failed initially -- this is why the first error is logged as a warning.
2. Most probably the removal error is the same for all tries, so it does not make sense to log it N times, or aggregate it.
3. In case it ultimately fails, since we do not return the error to the caller, it needs to be logged at least.

In any case, this is better than the current code and addresses the TODO in it.

Reply:
Previously, the error was ignored and we never knew if there was any retries, or what the removal error was.

Now, we log the first and the last error, so we still know what is going on, without any excessive logging.

Reply:
I took a look at whatever the current kernel implements, and normally it can only return EBUSY, so we'll log EBUSY once or twice. I still want to log the first error to distinguish between the "success from the first try" and "succeeded but after a retry cases". Can certainly demote the warning to debug if you like.

@mrunalp @cyphar @crosbymichael PTAL